Feature/protobuf

kylycht opened the pull request.

kylycht committed:
Add protobuf encoded connection

- SubProtoEncoded accepts topic and callback
- PublishProtoRequest accepts topic, reply and value interface(Message)
- PublishProtoEncoded accepts topic and value interface(Message)

kylycht committed:
Add connection check before taking an action

tectiv3 commented on nats.go:
```diff
@@ -171,6 +177,43 @@ func (n *Client) SubEncoded(topic string, callback natsCB) error {
+	if !n.c.IsConnected() {
+		return fmt.Errorf("NATS client is not connected")
+	}
+	if err := n.protoCon.PublishRequest(topic, reply, v); err != nil {
```
I still don't like that you are using publishrequest instead of normal request.

tectiv3 commented on nats.go:
```diff
@@ -171,6 +177,43 @@ func (n *Client) SubEncoded(topic string, callback natsCB) error {
+	}
+	if sub, err := n.protoCon.QueueSubscribe(topic, "distributed-queue-proto-encoded", callback); err == nil {
+		if sub.IsValid() && !n.isSubscribed(topic) {
+			n.subs[topic] = sub
```
this thing.
We should keep encoded subscriptions and plain text separate don't you think?

kylycht committed:
Perform Request instead of Pub/Sub

kylycht committed:
Set nats.Conn to Client

- If there was a call to Client before it creates connection, it will
result in nil pointer reference causing panic
